pilot/pkg/leaderelection/leaderelection_test.go

```diff
@@ -127,6 +139,114 @@ func TestPrioritizedLeaderElection(t *testing.T) {
 	close(stop7)
 }
 
+func TestMulticlusterLeaderElection(t *testing.T) {
+	client := fake.NewSimpleClientset()
+	watcher := &fakeDefaultWatcher{}
+	// First remote pod becomes the leader
+	_, stop := createElectionMulticluster(t, "pod1", "", true, watcher, nil, false, true, client)
+	// A new local pod cannot become leader
+	_, stop2 := createElectionMulticluster(t, "pod2", "", false, watcher, nil, false, false, client)
+	// A new remote pod cannot become leader
+	_, stop3 := createElectionMulticluster(t, "pod3", "", true, watcher, nil, false, false, client)
+	close(stop3)
+	close(stop2)
+	close(stop)
+}
+
+func SimpleRevisionComparison(currentLeaderRevision string, l *LeaderElection) bool {
+	// Old key comparison impl for interoperablilty testing
+	defaultRevision := l.defaultWatcher.GetDefault()
+	return l.revision != currentLeaderRevision &&
+		// empty default revision indicates that there is no default set
+		defaultRevision != "" && defaultRevision == l.revision
+}
+
+func TestPrioritizedMulticlusterLeaderElection(t *testing.T) {
```

Comment:
I was thinking something like this:
```

type instance struct {
	revision string
	remote   bool
	comp     string
}

func (i instance) GetComp() LeaderComparison {
	switch i.comp {
	case "local":
		return LocationPrioritizedComparison
	case "simple":
		return SimpleRevisionComparison
	default:
		panic("unknown comparison type")
	}
}
// TestPrioritizationCycles
func TestPrioritizationCycles(t *testing.T) {
	cases := []instance{}
	for _, rev := range []string{"", "default", "not-default"} {
		for _, loc := range []bool{false, true} {
			for _, comp := range []string{"local", "simple"} {
				cases = append(cases, instance{
					revision: rev,
					remote:   loc,
					comp:     comp,
				})
			}
		}
	}
	for _, start := range cases {
		t.Run(fmt.Sprint(start), func(t *testing.T) {
			checkCycles(t, start, cases, nil)
		})
	}
}

func alreadyHit(cur instance, chain []instance) bool {
	for _, cc := range chain {
		if cur == cc {
			return true
		}
	}
	return false
}

func checkCycles(t *testing.T, start instance, cases []instance, chain []instance) {
	if alreadyHit(start, chain) {
		t.Fatalf("cycle on leader election: cur %v, chain %v", start ,chain)
	}
	for _, nextHop := range cases {
		next := LeaderElection{
			remote:         nextHop.remote,
			defaultWatcher: &fakeDefaultWatcher{defaultRevision: "default"},
			revision:       nextHop.revision,
		}
		if nextHop.GetComp()(start.revision, &next) {
			nc := append([]instance{}, chain...)
			nc = append(nc, start)
			checkCycles(t, nextHop, cases, nc)
		}
	}
}
```

Then there are no changes in the core code required and its ~instant instead of taking ~20s or so. This ensures there are no cases where we steal the lock in an endless loop

Reply:
@howardjohn Thanks for the "crystal clear" explanation of your suggestion (i.e., writing the code). Makes sense.

Since your test handles the cycle checking, but not that the prioritization works as expected, I've also put back my original two other tests that did some basic checking of the functionality to make sure that remote functionality is at least making it through the system. PTAL.